Update architecture guide (#4778)
magento/devdocs: guides/v2.3/architecture/archi_perspectives/webapi-vision.md

### High-level Architecture

The following image provides an overview of how Web API requests are processed.

![Web API request processing overview](https://devdocs.magento.com/guides/v2.3/common/images/coding-standards/webapi-request-processing-high-level-overview.png)

### Extension Scenarios

#### Declare a new REST and SOAP endpoint

Any [service contract](https://devdocs.magento.com/guides/v2.3/extension-dev-guide/service-contracts/service-contracts.html) can be [exposed as REST and SOAP](https://devdocs.magento.com/guides/v2.3/extension-dev-guide/service-contracts/service-to-web-service.html) endpoints via configuration in a module's `webapi.xml` file.

#### Declare a new GraphQL query/mutation

1. Create custom resolvers.
1. Declare a custom query/mutation and all necessary types in `schema.graphqls`.

If the target module is called `MyModule`, then create the resolvers and configuration files in a new module called `MyModuleGraphQl`.

See the [GraphQL documentation](https://devdocs.magento.com/guides/v2.3/graphql/index.html) for more information.

#### Add a custom authentication mechanism

> 1. Provide a custom implementation of `\Magento\Authorization\Model\UserContextInterface`, that verifies a user's identity using a  custom authentication mechanism.

Reviewer: ```suggestion
1. Provide a custom implementation of `\Magento\Authorization\Model\UserContextInterface` that verifies a user's identity using a  custom authentication mechanism.
```

1. Web API schema should be strictly typed. (All complex types should eventually be resolved to scalar types.)
1. Authentication parameters must be passed via headers.
1. Throttling must be configured by the system integrator. It is not supported by Magento
1. Internal server errors must be masked and never shown to the user in production mode. In developer mode, original exceptions must never be masked and should be displayed along with the related stacktrace.
1. Pagination must be supported by all list operations.

#### GraphQL

1. Unlimited nesting should be supported during requests for related entities. (For example, get Order => Order Items => Products => Related Products)
1. Field filtration must be performed with SQL queries. Do not filter on the application layer after you've fetched all possible fields.
1. Third-party customizations must be done separately for Service Contracts and for GraphQL
1. For modularity purposes, GraphQL configuration must be declared in a separate module. For example, to expose GraphQL for the module `MyModule`, you must create the `schema.graphqls` file in the `MyModuleGraphQl` module.

> 1. GraphQL is primarily designed for store-front one-page apps and mobile applications. It supports token and cookie authentication, as well as guest access to public queries

Reviewer: ```suggestion
1. GraphQL is primarily designed for store-front, one-page apps and mobile applications. It supports token and cookie authentication, as well as guest access to public queries
```

1. All queries must return the 200 HTTP status code. If an error occurs, return the error in the response body. A 500 status code is allowed when an exception occurs when generating a schema, but not during requests.
1. The Store code should be passed via headers.

#### REST

1. The resource URL should be versioned (for example: V1). The version must be specified in the following format: `V\d.+`
1. Resource names in a URL should be in plural form (for example: products, carts)
1. ID parameters for operations on entities should be part of the resource URL (for example: /V1/products/**:sku** )
1. POST should be used to create an entity.
   PUT should be used to update an entity.

>    PATCH should be used to patch (update only selected fields) an entity(s) update.

Reviewer: ```suggestion
   PATCH should be used to patch (update only selected fields) an update to an entity or entities.
```

> 1. REST is designed for system integrations, mobile app integrations, as well as for one-page apps. It supports tokens, cookies and OAuth 1.0 with token exchange. It also supports guest access to public resources.

Reviewer: ```suggestion
1. REST is designed for system integrations, mobile app integrations, as well as for one-page apps. It supports tokens, cookies, and OAuth 1.0 with token exchange. It also supports guest access to public resources.
```

1. Responses must return responses with standard [HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes).

> 1. The store code must be passed via URL. For example `GET /rest/frenchStoreView/V1/products`. Persistence operations that should be performed for all stores at once, should have 'all' store code in the URL.

Reviewer: ```suggestion
1. The store code must be passed with a URL. For example `GET /rest/frenchStoreView/V1/products`. Persistence operations that should be performed for all stores at once should have 'all' store code in the URL.
```
